includes/class-sensei-course-enrolment-manager.php

```diff
@@ -190,4 +190,32 @@ public static function trigger_course_enrolment_check( $user_id, $course_id ) {
 			$course_enrolment->is_enrolled( $user_id, false );
 		}
 	}
+
+	/**
+	 * Recalculate all enrolments for a specific user. This method will return the cached enrolments if they already
+	 * exist. To enforce a calculation after a possible change, use
+	 * Sensei_Course_Enrolment_Manager::trigger_course_enrolment_check instead.
+	 *
+	 * @param int $user_id   User ID.
+	 *
+	 * @see Sensei_Course_Enrolment_Manager::trigger_course_enrolment_check
+	 */
+	public function recalculate_enrolments( $user_id ) {
```

Comment:
Since we'll call this method before _My Courses_ and _Learner Profile_, I wonder about moving `LEARNER_CALCULATION_META_NAME`, subsequent checking and recording, and maybe even the actual `get_enrolment_calculation_version` method to this class. As `Sensei_Enrolment_Calculation_Scheduler` already depends on this class, it shouldn't create a dependency loop.

Reply:
Yep. These also fit better in the manager class. I'll refactor.

Reply:
I was also wondering about doing the check here for `LEARNER_CALCULATION_META_NAME` so its state is tied to the user across the site and not just in the async provider.

Reply:
Ah sorry I missed that. I'll add the check there too. I'll keep it in the user query though since it simplifies batching.

Reply:
Cool. Yeah, definitely makes sense to keep the user query there. Looks like you were able to get around the bug you reported to core?

Reply:
Fixed in 10b7b69f529b2a64714b50053046a95448d99526

Reply:
Yeap! Instead of using a `NOT EXISTS` in `compare_key` I used it with `compare`. When you use it this way instead of doing a `WHERE NOT EXISTS` with a subquery, the query that is generated does a left join like this:
```
FROM
	wp_users u
	LEFT JOIN wp_usermmeta m,
		ON u.id = m.user_id AND m.meta_key = 'learner_calculated_version'
WHERE
	m.user IS NULL
```
Not sure if it makes any difference in performance but it gets the job done :)